File: src/test/java/ukalus/container/SetterCollectorTest.java
package ukalus.container;

import junit.framework.TestCase;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Stream;

public class SetterCollectorTest extends TestCase {

  public void testCollectionSetters() throws SecurityException, NoSuchMethodException {
    Map<String, Method> expected = new HashMap<>();
    expected.put("x", TestClass.class.getMethod("setX", Object.class));
    expected.put("someThingElse", TestClass.class.getMethod("setSomeThingElse", Object.class));
    Map<String, Method> actual = new HashMap<>();
    Stream.of(TestClass.class.getMethods()).forEach(new SetterCollector(actual));
    assertEquals(expected, actual);
  }

  class TestClass {

    public void setX(Object value) {

    }

    public void setSomeThingElse(Object value) {

    }

    public void setInvalid(Object value, Object value2) {

    }

    public void nonSetter(Object value) {

    }
  }

}

File: src/test/java/ukalus/container/SetterCollectorTest.kt
package ukalus.container

import junit.framework.TestCase
import java.lang.reflect.Method
import java.util.*

class SetterCollectorTest : TestCase() {

    fun testCollectionSetters() {
        val expected = HashMap<String, Method>()
        expected.put("x", TestClass::class.java.getMethod("setX", Any::class.java))
        expected.put("someThingElse", TestClass::class.java.getMethod("setSomeThingElse", Any::class.java))
        val actual = TestClass::class.java.methods
                .filter { it.name.startsWith("set") && it.parameterCount == 1 }
                .associateBy { Character.toLowerCase(it.name[3]) + it.name.substring(4) }
        TestCase.assertEquals(expected, actual)
    }

    internal inner class TestClass {

        fun setX(value: Any) {

        }

        fun setSomeThingElse(value: Any) {

        }

        fun setInvalid(value: Any, value2: Any) {

        }

        fun nonSetter(value: Any) {

        }
    }

}
